packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -650,8 +650,22 @@ class _SliverAppBarDelegate extends SliverPersistentHeaderDelegate {
   @override
   Widget build(BuildContext context, double shrinkOffset, bool overlapsContent) {
     final double visibleMainHeight = maxExtent - shrinkOffset - topPadding;
-    final double toolbarOpacity = pinned ? 1.0
-      : ((visibleMainHeight - _bottomHeight) / kToolbarHeight).clamp(0.0, 1.0);
+
+    // Truth table for `toolbarOpacity`:
```

Comment:
It looks like the result is fade when pinned=false and 1.0 when pinned=true except with one exception (the last row). Aha: I see that the implementation boils it down nicely:
```
!pinned || (floating && bottom != null)
```